### PR TITLE
feat: sync PRs with new format

### PR DESCRIPTION
**Pull request by minzcmu**
- sync PRs with new format, e.g. `PR-1:main`, `PR-1:second`
- store `workflowGraph` in pipeline

Related to https://github.com/screwdriver-cd/screwdriver/issues/723

### CODE REVIEW

File: lib/pipeline.js [hunk 1]
```diff
@@ -75,30 +81,32 @@ class PipelineModel extends BaseModel {
...
+            matchedPRs.forEach((job) => {
                 if (job.archived) {
-                    jobList.toUnarchive.push(existingPRs[index]);
+                    jobList.toUnarchive.push(job);
                 }
+            });
+
+            // if opened PR is not in the list of existingPRs, create it
+            if (matchedPRs.length === 0) {
+                jobList.toCreate.push({ name, ref: openedPRsRef[i].ref });
```
Comment (tkyi) [inline lib/pipeline.js]: I don't think you need `.ref` here. `openedPRsRef` should already just contain `job.ref`.

File: lib/pipeline.js [hunk 2]
```diff
@@ -124,13 +132,12 @@ class PipelineModel extends BaseModel {
     }
 
     /**
-     * Go through the list of job names and create it
+     * Go through the list of pr names and create pr jobs
      * @method _createJob
```
Comment (tkyi) [inline lib/pipeline.js]: `_createPRJob`

File: lib/pipeline.js [hunk 3]
```diff
@@ -4,9 +4,15 @@
 
 const BaseModel = require('./base');
 const parser = require('screwdriver-config-parser');
+const workflowParser = require('screwdriver-workflow-parser');
 const hoek = require('hoek');
+const { PR_JOB_NAME } = require('screwdriver-data-schema').config.regex;
 const PAGINATE_PAGE = 1;
 const PAGINATE_COUNT = 50;
+const PR_NAME_GROUP = {
```
Comment (jithine) [inline lib/pipeline.js]: can you call this `REGEX_CAPTURING_GROUP` ? Took a while to figure out this maps to `regex` used here https://github.com/screwdriver-cd/data-schema/commit/2e8596e87415aea93cab7085c6b5764368e86e19#diff-77dea1a61b658aff6ad9cd0d2528d2daR26

File: lib/pipeline.js [hunk 4]
```diff
@@ -75,30 +81,32 @@ class PipelineModel extends BaseModel {
             toUnarchive: []
         };
         const existingPRs = existingJobs.filter(j => j.isPR());   // list of PRs according to SD
-        const existingPRsNames = existingPRs.map(j => j.name);
         const openedPRsNames = openedPRs.map(j => j.name);
         const openedPRsRef = openedPRs.map(j => j.ref);
 
         existingPRs.forEach((job) => {
+            // getting PR and number e.g. PR-1
+            const prName = job.name.match(PR_JOB_NAME)[PR_NAME_GROUP.pr];
```
Comment (jithine) [inline lib/pipeline.js]: abstract this behind a function `getNameFromJob(job, type)` where type can be `pr` or `job`
```
function getNameFromJob(job, type) {
    return job.name.match(PR_JOB_NAME)[REGEX_CAPTURING_GROUP[type]];
}
```

File: lib/pipeline.js [hunk 5]
```diff
@@ -139,15 +146,24 @@ class PipelineModel extends BaseModel {
...
-                };
+        const jobsToCreate = prList.map(pr =>
+            this.getConfiguration(pr.ref).then((config) => {
+                const prJobNames = workflowParser.getNextJobs(config.workflowGraph, {
+                    trigger: '~pr',
+                    prNum: pr.name.split('-')[1]
+                });
+
+                return Promise.all[prJobNames.map((name) => {
+                    const jobName = name.match(PR_JOB_NAME)[PR_NAME_GROUP.job];
```
Comment (jithine) [inline lib/pipeline.js]: abstract this behind a function `getNameFromJob(job, type)`  as I've described in previous comment.

File: lib/pipeline.js [hunk 6]
```diff
@@ -139,15 +146,24 @@ class PipelineModel extends BaseModel {
...
         const factory = JobFactory.getInstance();
 
-        const jobsToCreate = jobList.map(j =>
-            this.getConfiguration(j.ref).then((config) => {
-                const jobConfig = {
-                    pipelineId: this.id,
-                    name: j.name,
-                    permutations: config.jobs.main
-                };
+        const jobsToCreate = prList.map(pr =>
```
Comment (jithine) [inline lib/pipeline.js]: how is this flow triggered ? what is starting point ? is it via webhooks ?

File: lib/pipeline.js [hunk 7]
```diff
@@ -75,30 +81,32 @@ class PipelineModel extends BaseModel {
...
-
-            if (index < 0) {
-                // if opened PR is not in the list of existingPRs, create it
-                jobList.toCreate.push({ name, ref: openedPRsRef[i] });
-            } else {
-                const job = existingPRs[index];
+            const matchedPRs = existingPRs.filter(job => job.name.startsWith(name));
 
-                // if opened PR was previously archived, unarchive it
+            // if opened PR was previously archived, unarchive it
```
Comment (petey) [inline lib/pipeline.js:112]: What does it mean to archive/unarchive a job?

Comment (minzcmu) [reply lib/pipeline.js:112]: We will archive a PR job if the PR got closed. Then it won't show up when we list jobs from the API. e.g.`pipelines/1/jobs`

**Review by jithine — APPROVED** [review]

**Review by jithine — APPROVED** [review]

File: lib/pipeline.js [hunk 8]
```diff
@@ -75,30 +92,33 @@ class PipelineModel extends BaseModel {
...
 
         openedPRsNames.forEach((name, i) => {
-            const index = existingPRsNames.indexOf(name);
-
-            if (index < 0) {
-                // if opened PR is not in the list of existingPRs, create it
-                jobList.toCreate.push({ name, ref: openedPRsRef[i] });
-            } else {
-                const job = existingPRs[index];
+            const matchedPRs = existingPRs.filter(job => job.name.startsWith(name));
```
Comment (tkyi) [inline lib/pipeline.js:110]: 👍

**Review by tkyi — APPROVED** [review]

**Pull request merged**